src/backend/v4l2/encoder.rs

```diff
@@ -0,0 +1,1411 @@
+use v4l2r::device::queue::RequestBuffersError;
+use v4l2r::device::AllocatedQueue;
+use v4l2r::device::Device;
+use v4l2r::device::Stream;
+use v4l2r::device::TryDequeue;
+use v4l2r::ioctl;
+use v4l2r::ioctl::BufferFlags;
+use v4l2r::ioctl::EncoderCommand;
+use v4l2r::ioctl::StreamOnError;
+use v4l2r::memory::BufferHandles;
+use v4l2r::memory::DmaBufHandle;
+use v4l2r::memory::MmapHandle;
+use v4l2r::memory::PlaneHandle;
+use v4l2r::memory::PrimitiveBufferHandles;
+use v4l2r::memory::UserPtrHandle;
+use v4l2r::Format;
+use v4l2r::PixelFormat;
+use v4l2r::QueueDirection;
+use v4l2r::QueueType;
+
+use crate::backend::v4l2::controls::VideoBitrate;
+use crate::backend::v4l2::controls::VideoBitrateMode;
+use crate::backend::v4l2::controls::VideoConstantQuality;
+use crate::backend::v4l2::controls::VideoForceKeyFrame;
+use crate::backend::v4l2::controls::VideoHeaderMode;
+use crate::encoder::stateful::BackendOutput;
+use crate::encoder::stateful::BackendRequest;
+use crate::encoder::stateful::BackendRequestId;
+use crate::encoder::stateful::StatefulBackendError;
+use crate::encoder::stateful::StatefulBackendResult;
+use crate::encoder::stateful::StatefulVideoEncoderBackend;
+use crate::encoder::CodedBitstreamBuffer;
+use crate::encoder::EncodeError;
+use crate::encoder::EncodeResult;
+use crate::encoder::FrameMetadata;
+use crate::encoder::RateControl;
+use crate::encoder::Tunings;
+use crate::utils::DmabufFrame;
+use crate::utils::UserPtrFrame;
+use crate::Fourcc;
+use crate::Resolution;
+
+#[derive(Debug, Error)]
+pub enum InitializationError {
+    #[error("failed to create a CAPTURE queue: {0:?}")]
+    CaptureQueueCreate(CreateQueueError),
+
+    #[error("failed to create a OUTPUT queue: {0:?}")]
+    OutputQueueCreate(CreateQueueError),
+
+    #[error("failed to set format for CAPTURE: {0:?}")]
+    SetFormatCapture(ioctl::SFmtError),
+
+    #[error("failed to set format for OUTPUT: {0:?}")]
+    SetFormatOutput(ioctl::SFmtError),
+
+    #[error("failed to request CAPTURE buffers: {0:?}")]
+    RequestBufferCatpure(RequestBuffersError),
+
+    #[error("failed to request OUTPUT buffers: {0:?}")]
+    RequestBufferOutput(RequestBuffersError),
+
+    #[error("failed to stream on CAPTURE: {0:?}")]
+    StreamOnCapture(StreamOnError),
+
+    #[error("failed to stream on OUTPUT: {0:?}")]
+    StreamOnOutput(StreamOnError),
+
+    #[error(transparent)]
+    EncoderStart(#[from] ioctl::EncoderCmdError),
+
+    #[error(transparent)]
+    CreatePollerError(nix::Error),
+
+    #[error(transparent)]
+    SetSelection(ioctl::SSelectionError),
+}
+
+#[derive(Debug, Error)]
+pub enum UnsupportedError {
+    #[error("frame upscaling")]
+    FrameUpscaling,
+
+    #[error("buffer lacking TIMESTAMP_COPY flag")]
+    NoTimestampCopyFlag,
+}
+
+#[derive(Debug, Error)]
+pub enum BackendError {
+    #[error(transparent)]
+    Unsupported(UnsupportedError),
+
+    #[error(transparent)]
+    InitializationError(#[from] InitializationError),
```

Comment:
I think we want to split this variant into its own error type - since `create` only returns `InitializationError` (with one exception that we can add to the new type), it removes to need for the double `map_err` while also making our errors more local.

Reply:
I think it should be better now. PTAL 😄